ctrd/watch.go

```diff
@@ -98,23 +98,31 @@ func (w *watch) add(pack *containerPack) {
 			exitTime: status.ExitTime(),
 		}
 
+		// NOTE: cleanup action should be taken only once!
+		var cleanupOnce sync.Once
+		cleanupFunc := func() error {
+			cleanupOnce.Do(func() {
+				if _, err := pack.task.Delete(context.Background()); err != nil {
+					logrus.Errorf("failed to delete task, container id: %s: %v", pack.id, err)
+				}
+
+				if err := pack.container.Delete(context.Background()); err != nil {
+					logrus.Errorf("failed to delete container, container id: %s: %v", pack.id, err)
+				}
+			})
+			return nil
+		}
+
 		if !pack.skipStopHooks {
 			for _, hook := range w.hooks {
-				if err := hook(pack.id, msg); err != nil {
+				if err := hook(pack.id, msg, cleanupFunc); err != nil {
 					logrus.Errorf("failed to execute the exit hooks: %v", err)
 					break
 				}
 			}
 		}
+		cleanupFunc()
```

Comment:
`if !pack.skipStopHooks`, the `cleanupFunc` will be called twice?

Reply:
I seen, you use the `sync.Once` to control the `cleanup` function can only be called once.

I check the `sync.Once` feature source code : 
```
// Once is an object that will perform exactly one action.
type Once struct {
	m    Mutex
	done uint32
}

// Do calls the function f if and only if Do is being called for the
// first time for this instance of Once. In other words, given
// 	var once Once
// if once.Do(f) is called multiple times, only the first call will invoke f,
// even if f has a different value in each invocation. A new instance of
// Once is required for each function to execute.
//
// Do is intended for initialization that must be run exactly once. Since f
// is niladic, it may be necessary to use a function literal to capture the
// arguments to a function to be invoked by Do:
// 	config.once.Do(func() { config.init(filename) })
//
// Because no call to Do returns until the one call to f returns, if f causes
// Do to be called, it will deadlock.
//
// If f panics, Do considers it to have returned; future calls of Do return
// without calling f.
//
func (o *Once) Do(f func()) {
	if atomic.LoadUint32(&o.done) == 1 {
		return
	}
	// Slow-path.
	o.m.Lock()
	defer o.m.Unlock()
	if o.done == 0 {
		defer atomic.StoreUint32(&o.done, 1)
		f()
	}
}
```

This function just can be called only once,  am I right?

Reply:
The function `cleanupOnce.Do` could be called multi times. But the function `f` as the parameter of `Do` would be exactly called once. @HusterWan

Reply:
@zhuangqh I got you!